L1Trigger/Phase2L1ParticleFlow/plugins/L1NNTauProducer.cc

```diff
@@ -153,9 +192,159 @@ void L1NNTauProducer::fillDescriptions(edm::ConfigurationDescriptions& descripti
+  L1TauEmu::pt_t pt_tot = 0;
+  input2_t p1_tot = 0;
+  input2_t p1x_tot = 0;
+  input2_t p1y_tot = 0;
+  input2_t p1z_tot = 0;
+
+  float p_tot = 0;
+  float px_tot = 0;
+  float py_tot = 0;
+  float pz_tot = 0;
+
+  float eta_1 = seed.eta();
+  float phi_1 = seed.phi();
+  input_t e1ta_1 = seed.eta();
+  input_t p1hi_1 = seed.phi();
+  L1TauEmu::pt_t pt = 0;
+  L1TauEmu::z0_t z0 = 0;
+  L1TauEmu::dxy_t dxy = 0;
+  for (unsigned i0 = 0; i0 < parts.size(); i0++) {
+    pt_tot = pt_tot + L1TauEmu::pt_t(parts[i0].pt());
+    if (L1TauEmu::inCone(seed, (parts[i0]), rCone2)) {
+      if (parts[i0].id() == l1t::PFCandidate::Electron || parts[i0].id() == l1t::PFCandidate::ChargedHadron ||
+          parts[i0].id() == l1t::PFCandidate::Photon) {
+        lId++;
+        pt = pt + L1TauEmu::pt_t(parts[i0].pt());
+        float deta = parts[i0].eta() - eta_1;
+        float dphi = parts[i0].phi() - phi_1;
+        float dr2 = deta * deta + dphi * dphi;
+        pz_tot = pz_tot + (parts[i0].pt()) * (1 - dr2 * 0.5);
+        py_tot = py_tot + (parts[i0].pt()) * dphi;  //sin(dphi ));
+        px_tot = px_tot + (parts[i0].pt()) * deta;  //sin(deta ));
+
+        input2_t d1eta = input_t(parts[i0].eta()) - e1ta_1;
+        input2_t d1phi = input_t(parts[i0].phi()) - p1hi_1;
+        input2_t d1r2 = d1eta * d1eta + d1phi * d1phi;
+        input2_t tmppt = input_t(parts[i0].pt());
+        input2_t half = 0.5;
+        p1z_tot = p1z_tot + tmppt * (1 - d1r2 * half);
+        p1y_tot = p1y_tot + tmppt * d1phi;
+        p1x_tot = p1x_tot + tmppt * d1eta;
+        p_tot = p_tot + (parts[i0].pt());
+        p1_tot = p1_tot + tmppt;
+        if (z0 == 0 && parts[i0].id() == l1t::PFCandidate::ChargedHadron) {
+          z0 = parts[i0].hwZ0();
+          dxy = parts[i0].hwDxy();
+        }
+      }
+    }
+  }
+  input2_t tmpmass1 = (p1_tot * p1_tot - p1x_tot * p1x_tot - p1y_tot * p1y_tot - p1z_tot * p1z_tot);
+  if (tmpmass1 < 0)
+    tmpmass1 = 0;
+  L1TauEmu::pt_t mass = l1ct::pt_t(tmpmass1);
+  if (pt < fSeedPt_)
+    return;
+  result_t NN = fTauNNIdHW_->compute(seed, parts);
+  input_t* lNNVector = fTauNNIdHW_->NNVectorVar();
+  float pNNVec[80];
+  for (unsigned i0 = 0; i0 < 80; i0++)
+    pNNVec[i0] = float(lNNVector[i0]);
+  L1TauEmu::etaphi_t eta = etaphi_t(seed.eta() * L1TauEmu::etaphi_base);
+  L1TauEmu::etaphi_t phi = etaphi_t(seed.phi() * L1TauEmu::etaphi_base);
+  math::PtEtaPhiMLorentzVector tempP4(
+      float(pt), float(eta) / L1TauEmu::etaphi_base, float(phi) / L1TauEmu::etaphi_base, float(mass));
+
+  //Make
+  l1t::PFTau l1PFTau(tempP4, pNNVec, NN, 0, lId);
+  l1PFTau.setZ0(float(z0) * 0.05);    //L1TauEmu::z0_base);
+  l1PFTau.setDxy(float(dxy) * 0.05);  //L1TauEmu::dxy_base);
+
+  //Firmware Tau
+  l1ct::Tau l1ctTau;
+  l1ctTau.hwPt = l1ct::pt_t(pt);  //l1gt is <16,11> and currently <16,14>
+  l1ctTau.hwEta = l1ct::eta_t(seed.eta() / l1ct::Scales::ETAPHI_LSB);
+  l1ctTau.hwPhi = l1ct::phi_t(seed.phi() / l1ct::Scales::ETAPHI_LSB);
```

Comment:
@Duchstf a bug has been spotted here, the eta/phi types should be `glbeta_t` and `glbphi_t`, as defined in the `l1ct::Tau object` [here](https://github.com/cms-l1t-offline/cmssw/blob/e6e8bdebe67709e85909347638ef5726250ded67/DataFormats/L1TParticleFlow/interface/taus.h#L16-L17).

Can you apply the fix?  Actually, it would be better to use [makeGlbPhi](https://github.com/cms-l1t-offline/cmssw/blob/e6e8bdebe67709e85909347638ef5726250ded67/DataFormats/L1TParticleFlow/interface/datatypes.h#L193) and [makeGlbEta](https://github.com/cms-l1t-offline/cmssw/blob/e6e8bdebe67709e85909347638ef5726250ded67/DataFormats/L1TParticleFlow/interface/datatypes.h#L190) instead?